### PR TITLE
Add contents and exercises

### LINKED ISSUE #22: provided solution doesn't match instructions - all in one file
Based on feedback from #17 

+ move content of `handler.solution.js` into multiple files
+ create tests

@eliascodes If it's ok with you, I think @majakudlicka and I might as well work on this, since some of the work has already been done via #16 and the extra sessions I've run with FAC11

### LINKED ISSUE #27: add a contents section at the top?
The readme is very long and a contents section would allow easier overview and navigation.

### PR DESCRIPTION
**Pull request by eliascodes**
### Fixes
fixes #27 
fixes #22

### Related to
related to #28 #29 #17 

### Summary
* Added contents section
* Added exercises to the end of each "approach" section so students can practise each approach

### Notes
They way I'm imagining this now, students needn't fully complete each section's exercise. The purpose is primarily illustrative and to provide more concrete examples. The test files are also part of the illustration; they provide examples for how to test using each kind of approach. 

Mentors should use their judgement, given their knowledge of the cohort, when deciding how much time to devote to these exercises. It may be that they decide to spend the whole workshop on these exercises and miss out the final exercise. Or it may by that they decide to skip quickly through the exercises, pointing out the solutions as resources that can be referred to while attempting the final exercise. Either approach could be appropriate depending on the level of the cohort.

### Review
I've requested reviews from everyone who's assigned to the `week-node-2` epic in master-reference. Any further comments welcome, in particular on the level of difficulty of the new exercises (though bearing in mind what I said above in **Notes**). I will merge once I have two approvals.

### CODE REVIEW

**Comment by eliascodes** [comment]
Hi week 5 people; week 5 in London is one week away. Please review.

**Comment by mattlub** [comment]
@eliascodes I've started looking at it, haven't look at all of it.

It does seem very very long now, I've seen your notes above but personally I'm not a fan of this sort of approach, where you have to pick and choose parts to do.

I'll finish reading it but also am happy to see it merged and see what the students make of it when it's delivered this time around.

**Comment by mattlub** [comment]
Also, the exercise tests should have messages... failing tests right now don't give any information, I'm getting 'test failed- should be truthy' :grin:

**Comment by eliascodes** [comment]
@mattlub thanks for taking a look, will fix the test messages, and add test commands to the README.

Would it still seem too long if you just ignored the final exercise? @polyccon also suggested splitting up the README into 3+ separate files, which might help from a purely psychological perspective.

I'd like to understand the issues you see in the 'pick and choose' approach?

**Comment by eliascodes** [comment]
tbh part of the problem with this workshop is that we tried to use it to cover server-side validation as well as error handling, which is probably not ideal. It's the reason that the final exercise is (almost necessarily) quite complex.

What if we just drop that entirely? At most there could then just be a much simpler (and _optional_) final exercise.

**Comment by polyccon** [comment]
I agree with @mattlub for his comments, I would also suggest commenting out all the tests, and in the testing instructions, say also only uncomment the tests relevant to each function they are currently working on so that they don't get confused with all the errors? I also feel that the 2nd sentence for each exercise might a bit redundant as it is repeating the first one..What do you think @eliascodes? Would you like me to do any of it and split the readme in smaller chunks as well?

**Comment by polyccon** [comment]
Yes perhaps server side validation could be moved in a separate workshop and we can allocate different time for that.

**Comment by eliascodes** [comment]
@polyccon sure, please go ahead and split up the README, and archive the server-side validation exercise in a separate folder.

**Comment by finnhodgkin** [comment]
Sorry I've been so, so, __so__ bad with week 5 stuff. I'm really glad other people have picked up the slack, especially since I've been in Naz.

I've just run through the workshop and have a couple questions,really nothing that I'd consider a blocker:

> Throwing should **only** be used in synchronous functions, **not** in asynchronous functions. Errors thrown from asynchronous functions will not be caught. To understand why, learn about the javascript [call stack](https://www.youtube.com/watch?v=8aGhZQkoFbQ).

1. While in general I agree with the above statement, putting so much stress on **only** and **not** implies that you literally _can't_ use try/catch near anything asynchronous _ever_. In reality I'd say it's less clear cut. Say you have an async operation and then in the callback you call a synchronous function, and that synchronous function has unreliable type handling that occasionally results in some kind of throw/error. At that point you would have a few options: 
  - take the time to carefully prep the data so the synchronous code won't error
  - rewrite the synchronous code
  - wrap the code in a try / catch _within_ the async function. 

It's not common and maybe I'm reading into the text too much, but I think the implication is that the following could _never_ be ok (I don't agree): 

```js
function waitAndRun(input, callback){
  setTimeout(function() {
    try {
      var yay = unreliableSynchronousFunction(input);
      callback(null, yay);
    } catch{
      callback('uhoh');
    }
  }, 1000);
}
``` 

Something like 'It's not possible to wrap an asynchronous function in a try/catch block, so throwing should **only** be used with synchronous code.' or 'Uncaught errors within asynchronous functions cannot be handled by the caller' would be more appropriate IMO

2. Do we have any workshops that do go over server side validation? Apart from research afternoon I don't remember anything in my cohort outside of the hapi weeks (Joi :heart:). If there's nothing then maybe dropping that section entirely would be harmful. Then again I guess having something at the end of a workshop that most people are unlikely to reach is basically the same thing.

**Comment by eliascodes** [comment]
@finnhodgkin thanks a lot for taking a look!

I agree that people may come away with the impression that your given example is bad practise, and that we shouldn't be giving them that impression. I will modify the text as per your suggestion.

> Do we have any workshops that do go over server side validation?

Not really, this was the only attempt to bring it in, and I think we can agree that it largely failed at that.

> Then again I guess having something at the end of a workshop that most people are unlikely to reach is basically the same thing.

Agree. I don't see the problem with putting it in a folder for now. If we can find space in the curriculum we can cannibalise parts of it into a more focused resource.

File: README.md [hunk 1]
```diff
@@ -230,6 +239,9 @@ Invalid argument: Second argument 2.3 is not an integer
   * This is actually a good thing, in light of [this](#kinds-of-errors) and [[8]](#resources).
```
Comment (finnhodgkin) [inline README.md]: Also this link doesn't work plus there's one more references to [8] that needs to be fixed. #resources doesn't exist anymore, it's #external-resources instead

**Comment by eliascodes** [comment]
@mattlub @finnhodgkin @polyccon I hope to have addressed all your comments now. Please take another quick look.

**Comment by mattlub** [comment]
@eliascodes haven't looked at everything, but I'm approving since you've addressed my comments.

**Pull request merged**